File: src/core/gui/Menu.ts
import { disposable_listener, el } from "./dom";
import { Widget } from "./Widget";
import { Property } from "../observable/property/Property";
import { property } from "../observable";
import { WritableProperty } from "../observable/property/WritableProperty";
import { WidgetProperty } from "../observable/property/WidgetProperty";
import "./Menu.css";

export class Menu<T> extends Widget {
    readonly element = el.div({ class: "core_Menu", tab_index: -1 });
    readonly selected: WritableProperty<T | undefined>;

    private readonly to_label: (element: T) => string;
    private readonly items: Property<readonly T[]>;
    private readonly inner_element = el.div({ class: "core_Menu_inner" });
    private readonly related_element: HTMLElement;
    private readonly _selected: WidgetProperty<T | undefined>;
    private hovered_index?: number;
    private hovered_element?: HTMLElement;

    constructor(
        items: readonly T[] | Property<readonly T[]>,
        to_label: (element: T) => string,
        related_element: HTMLElement,
    ) {
        super();

        this.visible.val = false;

        this.element.onmouseup = this.mouseup;
        this.element.onkeydown = this.keydown;

        this.inner_element.onmouseover = this.inner_mouseover;
        this.element.append(this.inner_element);

        this.to_label = to_label;
        this.items = Array.isArray(items) ? property(items) : (items as Property<readonly T[]>);
        this.related_element = related_element;

        this._selected = new WidgetProperty<T | undefined>(this, undefined, this.set_selected);
        this.selected = this._selected;

        this.disposables(
            this.items.observe(
                ({ value: items }) => {
                    this.inner_element.innerHTML = "";
                    this.inner_element.append(
                        ...items.map((item, index) =>
                            el.div({ text: to_label(item), data: { index: index.toString() } }),
                        ),
                    );
                    this.hover_item();
                },
                { call_now: true },
            ),

            disposable_listener(document, "mousedown", this.document_mousedown, {
                capture: true,
            }),

            disposable_listener(document, "keydown", this.document_keydown),
        );

        this.finalize_construction(Menu.prototype);
    }

    hover_next(): void {
        this.visible.set_val(true, { silent: false });
        this.hover_item(
            this.hovered_index != undefined ? (this.hovered_index + 1) % this.items.val.length : 0,
        );
    }

    hover_prev(): void {
        this.visible.set_val(true, { silent: false });
        this.hover_item(this.hovered_index ? this.hovered_index - 1 : this.items.val.length - 1);
    }

    select_hovered(): void {
        if (this.hovered_index != undefined) {
            this.select_item(this.hovered_index);
        }
    }

    protected set_visible(visible: boolean): void {
        super.set_visible(visible);

        if (this.visible.val != visible) {
            this.hover_item();
            this.inner_element.scrollTop = 0;
        }
    }

    protected set_selected(): void {
        // Noop
    }

    private mouseup = (e: Event): void => {
        if (!(e.target instanceof HTMLElement)) return;

        const index_str = e.target.dataset.index;
        if (index_str == undefined) return;

        this.select_item(parseInt(index_str, 10));
    };

    private keydown = (e: Event): void => {
        const key = (e as KeyboardEvent).key;

        switch (key) {
            case "ArrowDown":
                this.hover_next();
                break;

            case "ArrowUp":
                this.hover_prev();
                break;

            case "Enter":
                this.select_hovered();
                break;
        }
    };

    private inner_mouseover = (e: Event): void => {
        if (e.target && e.target instanceof HTMLElement) {
            const index = e.target.dataset.index;

            if (index != undefined) {
                this.hover_item(parseInt(index, 10));
            }
        }
    };

    private document_mousedown = (e: Event): void => {
        if (
            this.visible.val &&
            !this.element.contains(e.target as Node) &&
            !this.related_element.contains(e.target as Node)
        ) {
            this.visible.set_val(false, { silent: false });
        }
    };

    private document_keydown = (e: Event): void => {
        if ((e as KeyboardEvent).key === "Escape") {
            this.visible.set_val(false, { silent: false });
        }
    };

    private hover_item(index?: number): void {
        if (this.hovered_element) {
            this.hovered_element.classList.remove("core_Menu_hovered");
        }

        if (index == undefined) {
            this.hovered_index = undefined;
            this.hovered_element = undefined;
        } else {
            this.hovered_element = this.inner_element.children.item(index) as HTMLElement;

            if (this.hovered_element) {
                this.hovered_index = index;
                this.hovered_element.classList.add("core_Menu_hovered");
                this.hovered_element.scrollIntoView({ block: "nearest" });
            }
        }
    }

    private select_item(index: number): void {
        const item = this.items.val[index];
        if (item === undefined) return;

        this.selected.set_val(item, { silent: false });
        this.visible.set_val(false, { silent: false });
    }
}

File: webui/src/main/kotlin/world/phantasmal/webui/widgets/Menu.kt
package world.phantasmal.webui.widgets

import kotlinx.browser.document
import kotlinx.coroutines.CoroutineScope
import org.w3c.dom.*
import org.w3c.dom.events.Event
import org.w3c.dom.events.KeyboardEvent
import org.w3c.dom.events.MouseEvent
import world.phantasmal.observable.value.Val
import world.phantasmal.observable.value.falseVal
import world.phantasmal.observable.value.value
import world.phantasmal.webui.dom.disposableListener
import world.phantasmal.webui.dom.div
import world.phantasmal.webui.newJsObject

class Menu<T : Any>(
    scope: CoroutineScope,
    hidden: Val<Boolean> = falseVal(),
    disabled: Val<Boolean> = falseVal(),
    tooltip: String? = null,
    items: List<T>? = null,
    itemsVal: Val<List<T>>? = null,
    private val itemToString: (T) -> String = Any::toString,
    private val onSelect: (T) -> Unit = {},
    private val onCancel: () -> Unit = {},
) : Widget(
    scope,
    hidden,
    disabled,
    tooltip,
) {
    private val items: Val<List<T>> = itemsVal ?: value(items ?: emptyList())
    private lateinit var innerElement: HTMLElement
    private var highlightedIndex: Int? = null
    private var highlightedElement: Element? = null
    private var previouslyFocusedElement: Element? = null

    override fun Node.createElement() =
        div {
            className = "pw-menu"
            tabIndex = -1
            onmouseup = ::onMouseUp
            onkeydown = ::onKeyDown
            onblur = { onBlur() }

            innerElement = div {
                className = "pw-menu-inner"
                onmouseover = ::innerMouseOver

                bindChildrenTo(items) { item, index ->
                    div {
                        dataset["index"] = index.toString()
                        textContent = itemToString(item)
                    }
                }
            }

            observe(this@Menu.hidden) {
                if (it) {
                    document.removeEventListener("mousedown", ::onDocumentMouseDown)
                    clearHighlightItem()

                    (previouslyFocusedElement as HTMLElement?)?.focus()
                } else {
                    document.addEventListener("mousedown", ::onDocumentMouseDown)
                }
            }

            observe(disabled) {
                if (it) {
                    clearHighlightItem()
                }
            }

            disposableListener(document, "keydown", ::onDocumentKeyDown)
        }

    override fun internalDispose() {
        document.removeEventListener("mousedown", ::onDocumentMouseDown)
        super.internalDispose()
    }

    override fun focus() {
        previouslyFocusedElement = document.activeElement
        super.focus()
    }

    fun highlightItem(item: T) {
        val idx = items.value.indexOf(item)

        if (idx != -1) {
            highlightItemAt(idx)
        }
    }

    private fun onMouseUp(e: MouseEvent) {
        val target = e.target

        if (target !is HTMLElement) return

        target.dataset["index"]?.toIntOrNull()?.let(::selectItem)
    }

    private fun onKeyDown(e: KeyboardEvent) {
        when (e.key) {
            "ArrowDown" -> {
                e.preventDefault()
                highlightItemAt(
                    when (val idx = highlightedIndex) {
                        null, items.value.lastIndex -> 0
                        else -> idx + 1
                    }
                )
            }

            "ArrowUp" -> {
                e.preventDefault()
                highlightItemAt(
                    when (val idx = highlightedIndex) {
                        null, 0 -> items.value.lastIndex
                        else -> idx - 1
                    }
                )
            }

            "Enter", " " -> {
                e.preventDefault()
                e.stopPropagation()
                highlightedIndex?.let(::selectItem)
            }
        }
    }

    private fun onBlur() {
        onCancel()
    }

    private fun innerMouseOver(e: MouseEvent) {
        val target = e.target

        if (target is HTMLElement) {
            target.dataset["index"]?.toIntOrNull()?.let(::highlightItemAt)
        }
    }

    private fun onDocumentMouseDown(e: Event) {
        val target = e.target

        if (target !is Node || !element.contains(target)) {
            onCancel()
        }
    }

    private fun onDocumentKeyDown(e: KeyboardEvent) {
        if (e.key === "Escape") {
            onCancel()
        }
    }

    private fun clearHighlightItem() {
        highlightedElement?.classList?.remove("pw-menu-highlighted")
        highlightedIndex = null
        highlightedElement = null
    }

    private fun highlightItemAt(index: Int) {
        highlightedElement?.classList?.remove("pw-menu-highlighted")

        if (disabled.value) return

        highlightedElement = innerElement.children.item(index)

        highlightedElement?.let {
            highlightedIndex = index
            it.classList.add("pw-menu-highlighted")
            it.scrollIntoView(newJsObject { block = "nearest" })
        }
    }

    private fun selectItem(index: Int) {
        if (disabled.value) return

        items.value.getOrNull(index)?.let(onSelect)
    }

    companion object {
        init {
            @Suppress("CssUnusedSymbol", "CssUnresolvedCustomProperty")
            // language=css
            style("""
                .pw-menu {
                    z-index: 1000;
                    position: absolute;
                    box-sizing: border-box;
                    outline: none;
                    border: var(--pw-control-border);
                    --scrollbar-color: hsl(0, 0%, 18%);
                    --scrollbar-thumb-color: hsl(0, 0%, 22%);
                }

                .pw-menu > .pw-menu-inner {
                    overflow: auto;
                    background-color: var(--pw-control-bg-color);
                    max-height: 500px;
                    border: var(--pw-control-inner-border);
                }

                .pw-menu > .pw-menu-inner > * {
                    padding: 4px 8px;
                    white-space: nowrap;
                }

                .pw-menu > .pw-menu-inner > .pw-menu-highlighted {
                    background-color: var(--pw-control-bg-color-hover);
                    color: var(--pw-control-text-color-hover);
                }
            """.trimIndent())
        }
    }
}
